### PR TITLE
Implement shows across a whole pyramid side

### PR DESCRIPTION
**Pull request by silverlyra**
This change gives shows a `Grid` spanning multiple panels.

IMG_0904

### Breaking change:

**Run `go_tri` with `--layout one` or `--layout two` to drive one or two panels.** By default it'll assume you're driving the whole pyramid, and crash because that layout is not actually defined yet.

### Code changes:

- Cell ID's are gone. `Coordinate` is now the native cell location type.
- `Grid` is now an abstract class, implemented by `Face`, which can run a single panel or a whole side of the pyramid.
- `Face`s are constructed from a list of `Panel`s, which give their origin (bottom-left) coordinate and their start DMX address.
- `Address`, `Coordinate`, etc. were moved to `grid.geom`.
- `Universe` is now a NamedTuple, which stores the actual universe ID as well as the ID of the first universe of its panel (so we can tell which universes are shorter).

### CODE REVIEW

File: grid/grid.py [hunk 1]
```diff
@@ -74,53 +86,46 @@ def pixels(self, sel: Selector, direction: Direction = Direction.NATURAL) -> Ite
...
-            cell_id = loc.pos(self.geom).id
-        else:
-            cell_id = loc
+        coordinate = self._normalize_location(loc)
+        cell = self._cell(coordinate)
 
-        if cell_id < 0:
-            raise KeyError(cell_id)
+        if cell is None:
+            # if coordinate not in self.geom:
```
Comment (silverlyra) [inline grid/grid.py]: this needs cleanup

File: grid/cell.py [hunk 2]
```diff
@@ -34,19 +33,28 @@ class Cell(NamedTuple):
     the panels.
     """
 
-    position: "Position"
+    coordinate: Coordinate
     orientation: Orientation
     addresses: List["Address"]
     geom: Geometry
+    real: bool
```
Comment (sul3n3t) [inline grid/cell.py]: Could be an enum, cell in a panel, cell in the blank space of the face, and cell off the edge of the face.

File: go_tri.py [hunk 3]
```diff
@@ -326,7 +358,12 @@ def go_web(self):
             [name for (name, cls) in shows.load_shows()]))
         sys.exit(0)
 
+    grid = build_grid(args.layout)
+
     if args.simulator:
+        if args.layout != "one":
```
Comment (sul3n3t) [inline go_tri.py]: Can `--simulator` imply `layout=one` for now?

File: go_tri.py [hunk 4]
```diff
@@ -297,15 +298,46 @@ def go_web(self):
...
+
+    if layout == 'one':
+        geom = Geometry(origin=Coordinate(0, 0), rows=ROWS_PER_PANEL)
+        panels = [
+            Panel(
+                Geometry(origin=Coordinate(0, 0), rows=ROWS_PER_PANEL),
+                Address(Universe(1, 1), 4),
+            ),
+        ]
+    elif layout == 'two':
```
Comment (sul3n3t) [inline go_tri.py]: This is simulating the `Face` as

```
   T_fake
T_one  T_two
```

right?

File: grid/face.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+PIXELS_IN_ROW_TURNAROUND: int = 11
+
+
+class Panel(NamedTuple):
+    """
+    Panel represents an assembled panel in the pyramid frame.
+    """
+
+    geom: Geometry
+    start: Address
```
Comment (sul3n3t) [inline grid/face.py]: Doesn't have to be this PR, but is this a good place to configure the dead pixel to skip on the one panel?

File: model/base.py [hunk 6]
```diff
@@ -14,13 +14,29 @@ def set(self, cell: Cell, addr: Address, color: Color):
 
         addr is an Address, except in the case of the simulator it is a cell ID.
         """
+        raise NotImplementedError
```
Comment (sul3n3t) [inline model/base.py]: Good catch

File: grid/face.py (hunk 5 above)
Comment (silverlyra) [reply grid/face.py]: yeah, we def could

File: grid/cell.py (hunk 2 above)
Comment (silverlyra) [reply grid/cell.py]: > off the edge of the face

I put back the `coordinate not in self.geom` check, so if you try to access a cell that's not even defined in the whole `Face` geometry, it's undefined. `Cell`s with `real=false` will only be emitted for the gaps between panels

**Pull request merged**